azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/service/CallHistoryService.kt

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.android.communication.ui.calling.service
+
+import com.azure.android.communication.ui.calling.data.CallHistoryRepository
+import com.azure.android.communication.ui.calling.redux.Store
+import com.azure.android.communication.ui.calling.redux.state.ReduxState
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+import org.threeten.bp.LocalDateTime
+
+internal interface CallHistoryService {
+    fun start(coroutineScope: CoroutineScope)
+}
+
+internal class CallHistoryServiceImpl(
+    private val store: Store<ReduxState>,
+    private val callHistoryRepository: CallHistoryRepository,
+) : CallHistoryService {
+    private val callIdStateFlow = MutableStateFlow<String?>(null)
+
+    override fun start(coroutineScope: CoroutineScope) {
+        coroutineScope.launch {
+            store.getStateFlow().collect {
+                callIdStateFlow.value = it.callState.callId
+            }
+        }
+
+        coroutineScope.launch {
+            callIdStateFlow.collect { callId ->
+                val callStartDateTime = store.getCurrentState().callState.callStartLocalDateTime
+                if (callId != null && callStartDateTime != null) {
+                    callHistoryRepository.insert(callId, callStartDateTime)
+                }
+            }
+        }
+
+        coroutineScope.launch {
+            cleanOldRecords()
```

Comment:
I guess it's fine to do the cleanup on every insert, and not just once during `init`.

But, if you're going to do this, I would just hide the cleanup login as part of the insert logic within the repository implementation, and do both INSERT and DELETE wrapped together in a single transaction.

It'll be faster and more robust. Less transaction handling for sqlite to do with a single transaction; as-is, you're implicitly beginning and ending at least 2+N transactions (N=# of records to cleanup).

Reply:
Moved.